File: PoketDex/app/src/main/java/pt/rfsfernandes/data/local/PokemonTypeConverters.java
package pt.rfsfernandes.data.local;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.Collections;
import java.util.List;

import androidx.room.TypeConverter;
import pt.rfsfernandes.model.SimpleModelData;
import pt.rfsfernandes.model.pokemon.abilities.Ability;
import pt.rfsfernandes.model.pokemon.moves.PokemonMoves;
import pt.rfsfernandes.model.pokemon.sprites.OfficialArtWork;
import pt.rfsfernandes.model.pokemon.sprites.OtherSpriteInfo;
import pt.rfsfernandes.model.pokemon.sprites.PokemonSprites;
import pt.rfsfernandes.model.pokemon.stats.Stats;
import pt.rfsfernandes.model.pokemon.types.PokemonType;
import pt.rfsfernandes.model.service_responses.PokemonResult;

public class PokemonTypeConverters {
  // Pokemon Result
  @TypeConverter
  public static List<PokemonResult> stringToPokemonResultList(String data) {
    if (data == null) {
      return Collections.emptyList();
    }

    Type listType = new TypeToken<List<PokemonResult>>() {
    }.getType();

    return new Gson().fromJson(data, listType);
  }

  @TypeConverter
  public static String pokemonResultToString(List<PokemonResult> someObjects) {
    return new Gson().toJson(someObjects);
  }

  // SimpleModelData
  @TypeConverter
  public static List<SimpleModelData> stringToSimpleModelDataList(String data) {
    if (data == null) {
      return Collections.emptyList();
    }

    Type listType = new TypeToken<List<SimpleModelData>>() {
    }.getType();

    return new Gson().fromJson(data, listType);
  }

  @TypeConverter
  public static String simpleModelDataListToString(List<SimpleModelData> someObjects) {
    return new Gson().toJson(someObjects);
  }

  @TypeConverter
  public static SimpleModelData stringToSimpleModelData(String data) {
    if (data == null) {
      return new SimpleModelData("", "");
    }

    Type listType = new TypeToken<SimpleModelData>() {
    }.getType();

    return new Gson().fromJson(data, listType);
  }

  @TypeConverter
  public static String simpleModelDataToString(SimpleModelData someObjects) {
    return new Gson().toJson(someObjects);
  }

  // Ability
  @TypeConverter
  public static List<Ability> stringToAbility(String data) {
    if (data == null) {
      return Collections.emptyList();
    }

    Type listType = new TypeToken<List<Ability>>() {
    }.getType();

    return new Gson().fromJson(data, listType);
  }

  @TypeConverter
  public static String abilityToString(List<Ability> someObjects) {
    return new Gson().toJson(someObjects);
  }

  // Moves
  @TypeConverter
  public static List<PokemonMoves> stringToMoves(String data) {
    if (data == null) {
      return Collections.emptyList();
    }

    Type listType = new TypeToken<List<PokemonMoves>>() {
    }.getType();

    return new Gson().fromJson(data, listType);
  }

  @TypeConverter
  public static String movesToString(List<PokemonMoves> someObjects) {
    return new Gson().toJson(someObjects);
  }

  // Stats
  @TypeConverter
  public static List<Stats> stringToStats(String data) {
    if (data == null) {
      return Collections.emptyList();
    }

    Type listType = new TypeToken<List<Stats>>() {
    }.getType();

    return new Gson().fromJson(data, listType);
  }

  @TypeConverter
  public static String statsToString(List<Stats> someObjects) {
    return new Gson().toJson(someObjects);
  }

  // PokemonType
  @TypeConverter
  public static List<PokemonType> stringToPokemonType(String data) {
    if (data == null) {
      return Collections.emptyList();
    }

    Type listType = new TypeToken<List<PokemonType>>() {
    }.getType();

    return new Gson().fromJson(data, listType);
  }

  @TypeConverter
  public static String pokemonTypeToString(List<PokemonType> someObjects) {
    return new Gson().toJson(someObjects);
  }

  // PokemonSprites
  @TypeConverter
  public static PokemonSprites stringToPokemonSprites(String data) {
    if (data == null) {
      return new PokemonSprites("", "", new OtherSpriteInfo(new OfficialArtWork("")));
    }

    Type listType = new TypeToken<PokemonSprites>() {
    }.getType();

    return new Gson().fromJson(data, listType);
  }

  @TypeConverter
  public static String pokemonSpritesToString(PokemonSprites someObjects) {
    return new Gson().toJson(someObjects);
  }

  // OtherSpriteInfo
  @TypeConverter
  public static List<OtherSpriteInfo> stringToOtherSpriteInfo(String data) {
    if (data == null) {
      return Collections.emptyList();
    }

    Type listType = new TypeToken<List<OtherSpriteInfo>>() {
    }.getType();

    return new Gson().fromJson(data, listType);
  }

  @TypeConverter
  public static String otherSpriteInfoToString(List<OtherSpriteInfo> someObjects) {
    return new Gson().toJson(someObjects);
  }

  // OfficialArtWork
  @TypeConverter
  public static List<OfficialArtWork> stringToOfficialArtWork(String data) {
    if (data == null) {
      return Collections.emptyList();
    }

    Type listType = new TypeToken<List<OfficialArtWork>>() {
    }.getType();

    return new Gson().fromJson(data, listType);
  }

  @TypeConverter
  public static String officialArtWorkToString(List<OfficialArtWork> someObjects) {
    return new Gson().toJson(someObjects);
  }


}

File: PoketDex/app/src/main/java/pt/rfsfernandes/pocketdex/data/local/PokemonTypeConverters.kt
package pt.rfsfernandes.pocketdex.data.local

import androidx.room.TypeConverter
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import pt.rfsfernandes.pocketdex.model.SimpleModelData
import pt.rfsfernandes.pocketdex.model.pokemon.moves.PokemonMoves
import pt.rfsfernandes.pocketdex.model.pokemon.sprites.OfficialArtWork
import pt.rfsfernandes.pocketdex.model.pokemon.sprites.OtherSpriteInfo
import pt.rfsfernandes.pocketdex.model.pokemon.sprites.PokemonSprites
import pt.rfsfernandes.pocketdex.model.pokemon.stats.Stats
import pt.rfsfernandes.pocketdex.model.pokemon.types.PokemonType
import pt.rfsfernandes.pocketdex.model.service_responses.PokemonResult
import pt.rfsfernandes.pocketdex.model.type.DamageRelations

object PokemonTypeConverters {
    // Pokemon Result
    @TypeConverter
    fun stringToPokemonResultList(data: String?): List<PokemonResult> {
        if (data == null) {
            return emptyList()
        }
        val listType = object : TypeToken<List<PokemonResult?>?>() {}.type
        return Gson().fromJson(data, listType)
    }

    @TypeConverter
    fun pokemonResultToString(someObjects: List<PokemonResult?>?): String {
        return Gson().toJson(someObjects)
    }

    // SimpleModelData
    @TypeConverter
    fun stringToSimpleModelDataList(data: String?): List<SimpleModelData> {
        if (data == null) {
            return emptyList()
        }
        val listType = object : TypeToken<List<SimpleModelData?>?>() {}.type
        return Gson().fromJson(data, listType)
    }

    @TypeConverter
    fun simpleModelDataListToString(someObjects: List<SimpleModelData?>?): String {
        return Gson().toJson(someObjects)
    }

    @TypeConverter
    fun stringToSimpleModelData(data: String?): SimpleModelData {
        if (data == null) {
            return SimpleModelData("", "")
        }
        val listType = object : TypeToken<SimpleModelData?>() {}.type
        return Gson().fromJson(data, listType)
    }

    @TypeConverter
    fun simpleModelDataToString(someObjects: SimpleModelData?): String {
        return Gson().toJson(someObjects)
    }

    // Moves
    @TypeConverter
    fun stringToMoves(data: String?): List<PokemonMoves> {
        if (data == null) {
            return emptyList()
        }
        val listType = object : TypeToken<List<PokemonMoves?>?>() {}.type
        return Gson().fromJson(data, listType)
    }

    @TypeConverter
    fun movesToString(someObjects: List<PokemonMoves?>?): String {
        return Gson().toJson(someObjects)
    }

    // Stats
    @TypeConverter
    fun stringToStats(data: String?): List<Stats> {
        if (data == null) {
            return emptyList()
        }
        val listType = object : TypeToken<List<Stats?>?>() {}.type
        return Gson().fromJson(data, listType)
    }

    @TypeConverter
    fun statsToString(someObjects: List<Stats?>?): String {
        return Gson().toJson(someObjects)
    }

    // PokemonType
    @TypeConverter
    fun stringToPokemonType(data: String?): List<PokemonType> {
        if (data == null) {
            return emptyList()
        }
        val listType = object : TypeToken<List<PokemonType?>?>() {}.type
        return Gson().fromJson(data, listType)
    }

    @TypeConverter
    fun pokemonTypeToString(someObjects: List<PokemonType?>?): String {
        return Gson().toJson(someObjects)
    }

    // PokemonSprites
    @TypeConverter
    fun stringToPokemonSprites(data: String?): PokemonSprites {
        if (data == null) {
            return PokemonSprites("", "", OtherSpriteInfo(OfficialArtWork("")))
        }
        val listType = object : TypeToken<PokemonSprites?>() {}.type
        return Gson().fromJson(data, listType)
    }

    @TypeConverter
    fun pokemonSpritesToString(someObjects: PokemonSprites?): String {
        return Gson().toJson(someObjects)
    }

    // OtherSpriteInfo
    @TypeConverter
    fun stringToOtherSpriteInfo(data: String?): List<OtherSpriteInfo> {
        if (data == null) {
            return emptyList()
        }
        val listType = object : TypeToken<List<OtherSpriteInfo?>?>() {}.type
        return Gson().fromJson(data, listType)
    }

    @TypeConverter
    fun otherSpriteInfoToString(someObjects: List<OtherSpriteInfo?>?): String {
        return Gson().toJson(someObjects)
    }

    // OfficialArtWork
    @TypeConverter
    fun stringToOfficialArtWork(data: String?): List<OfficialArtWork> {
        if (data == null) {
            return emptyList()
        }
        val listType = object : TypeToken<List<OfficialArtWork?>?>() {}.type
        return Gson().fromJson(data, listType)
    }

    @TypeConverter
    fun officialArtWorkToString(someObjects: List<OfficialArtWork?>?): String {
        return Gson().toJson(someObjects)
    }

    // PokemonSprites
    @TypeConverter
    fun stringToDamageRelations(data: String?): DamageRelations {
        if (data == null) {
            return DamageRelations(
                ArrayList(), ArrayList(), ArrayList(),
                ArrayList(), ArrayList(), ArrayList()
            )
        }
        val listType = object : TypeToken<DamageRelations?>() {}.type
        return Gson().fromJson(data, listType)
    }

    @TypeConverter
    fun damageRelationsToString(someObjects: DamageRelations?): String {
        return Gson().toJson(someObjects)
    }
}